torch/lib/THD/base/DataChannel.hpp

```diff
@@ -75,20 +75,52 @@ struct DataChannel {
 
   virtual bool init() = 0;
 
+  /**
+   * This is required for NCCL backend, since the destroy cannot be done before
+   * CUDA is unloaded since DataChannel is a static object.
+   */
+  virtual void destroy() = 0;
+
   virtual rank_type getRank() = 0;
   virtual rank_type getNumProcesses() = 0;
 
-  virtual void allGather(std::vector<at::Tensor>& output, at::Tensor& input,
+  // All gather multiple GPUs on a number of nodes
+  virtual void allGather(std::vector<at::Tensor>& input,
+                         std::vector<at::Tensor>& output,
```

Comment:
This signature is also weird. Firstly, it swaps output and input compared to the other overload. Secondly, I'd expect that if you get a vector of inputs, you also get a vector of outputs (a vector per device).